sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveUnion.scala

```diff
@@ -20,137 +20,66 @@ package org.apache.spark.sql.catalyst.analysis
-        val sorted = sortStructFieldsInWithFields(reducedStruct)
-        sorted
-      }.get
+    val colType = col.dataType.asInstanceOf[StructType]
+
+    val newStructFields = mutable.ArrayBuffer.empty[Expression]
+
+    val targetStructFields = targetType.fields.foreach { expectedField =>
+      val currentField = colType.fields.find(f => resolver(f.name, expectedField.name))
+
+      val newExpression = (currentField, expectedField.dataType) match {
+        case (Some(cf), expectedType: StructType) if cf.dataType.isInstanceOf[StructType] =>
+            val extractedValue = ExtractValue(col, Literal(cf.name), resolver)
+            val combinedStruct = addFields(extractedValue, expectedType)
+            if (extractedValue.nullable) {
+              If(IsNull(extractedValue),
+                Literal(null, combinedStruct.dataType),
+                combinedStruct)
+            } else {
+              combinedStruct
+            }
+        case (Some(cf), _) =>
+          ExtractValue(col, Literal(cf.name), resolver)
+        case (None, expectedType) =>
+          Literal(null, expectedType)
+      }
+      newStructFields ++= Literal(expectedField.name) :: newExpression :: Nil
     }
-  }
 
-  /**
-   * Adds missing fields recursively into given `col` expression. The missing fields are given
-   * in `fields`. For example, given `col` as "z struct<z:int, y:int>, x int", and `fields` is
-   * "z struct<w:long>, w string". This method will add a nested `z.w` field and a top-level
-   * `w` field to `col` and fill null values for them. Note that because we might also add missing
-   * fields at other side of Union, we must make sure corresponding attributes at two sides have
-   * same field order in structs, so when we adding missing fields, we will sort the fields based on
-   * field names. So the data type of returned expression will be
-   * "w string, x int, z struct<w:long, y:int, z:int>".
-   */
-  private def addFieldsInto(
-      col: Expression,
-      fields: Seq[StructField]): Expression = {
-    fields.foldLeft(col) { case (currCol, field) =>
-      field.dataType match {
-        case st: StructType =>
-          val resolver = conf.resolver
-          val colField = currCol.dataType.asInstanceOf[StructType]
-            .find(f => resolver(f.name, field.name))
-          if (colField.isEmpty) {
-            // The whole struct is missing. Add a null.
-            UpdateFields(currCol, field.name, Literal(null, st))
-          } else {
-            UpdateFields(currCol, field.name,
-              addFieldsInto(ExtractValue(currCol, Literal(field.name), resolver), st.fields))
-          }
-        case dt =>
-          UpdateFields(currCol, field.name, Literal(null, dt))
+    colType.fields
+      .filter(f => targetType.fields.find(tf => resolver(f.name, tf.name)).isEmpty)
+      .foreach { f =>
+        newStructFields ++= Literal(f.name) :: ExtractValue(col, Literal(f.name), resolver) :: Nil
```

Comment:
Is this to add fields only in left side at the end of struct? Doesn't it match original field order?

Reply:
When the left is projected this should match the original, but when the right is projected this will contain things on the right that aren't in the left. Basically it's

```
rightChild = left ++ (right - left)
leftChild = rightChild ++ (left - rightChild) = rightChild
```

Reply:
where is the project?


`newStructFields` contains the (nested) struct fields both in left and right column in right order.

Then here it adds (nested) struct fields only in left back to `newStructFields`, before create new struct (`CreateNamedStruct`).

Do we reorder the fields later?

Reply:
`rightProjectList` contains the nested structs mapped in the order of left fields then remaining right fields recursively, so that's where all the reordering happens

Reply:
And then leftChild is created from the fields in rightChild which already has all the fields as that point, which is the left fields and then the right fields

Reply:
> rightProjectList contains the nested structs mapped in the order of left fields then remaining right fields recursively, so that's where all the reordering happens

The projection projects original right attributes to `rightProjectList`. If you have different nested column order, it will be projected to new order.

Reply:
I look the code more in details.

`targetType` is actually left side type in first call. So here we align a right struct column to left struct column.

So it makes sense to add left nested columns first (`newStructFields`), then add nested columns only in right struct.

Reply:
Yeah I kept most of the naming which gets a little weird with how left/right/source/target are constructed and depends on where in the codepath you are